### PR TITLE
Milestone 1.1: Write test for contest.spec.ts ( part-2)

### PR DESCRIPTION
**Pull request by ShivamJhaa**
# Descripción

* Formatted the old code with eslint formatter (we are using eslint throughout the codebase)
* Added cypress test to give a past contest as a virtual contest
* Added cypress test to test practice mode in a contest
* Added cypress test to create a contest and review ranking contests when the scoreboard shows me has finished
* Refactored code to avoid redundancy (created new answerClarification function as it was getting used twice)

Proof:

https://github.com/omegaup/omegaup/assets/86483059/005ec7f8-a535-4a91-8dc5-0c30b99292d0

# Comentarios

Si hay comentarios para los reviewers (por ejemplo, dónde hay que empezar a
revisar, algo a lo que hay que poner atención adicional, etc.), todo eso se
puede poner en esta sección.

También si faltan cosas por hacer, se puede hacer un checklist en esta sección.

Si no se utiliza, esta sección se puede eliminar.

# Checklist:

- [ ] El código sigue la [guía de
      estilo](https://github.com/omegaup/omegaup/wiki/Coding-guidelines) de
      omegaUp.
- [ ] Se corrieron todas las pruebas y pasaron.
- [ ] Si se está agregando funcionalidad nueva, se agregaron pruebas.
- [ ] Si el cambio es grande (> 200 líneas), hay que intentar partirlo en
      varios pull requests. De preferencia uno para los controladores + phpunit
      y luego otro para la interfaz.

### CODE REVIEW

**Comment by ShivamJhaa** [comment]
The initial 3 commits are not related to this PR (they are present because I messed up changes on the main branch). So please do not review them.

**Comment by ShivamJhaa** [comment]
@heduenas @Mau-MD Please take a look.
Thanks!

**Comment by heduenas** [comment]
Can you please move away from using the `omegaup` user? (I missed this on your previous PR).
The `omegaup` user has site-admin permissions that the regular teacher-account doesn't have.

File: cypress/integration/contest.spec.ts [hunk 1]
```diff
@@ -105,16 +98,129 @@ describe('Contest Test', () => {
...
+
+  it('Should give a past contest as a virtual contest', () => {
+    const contestOptions = contestPage.generateContestOptions();
+    const now = new Date();
+    const userLoginOptions = loginPage.registerMultipleUsers(1);
+    const users = [userLoginOptions[0].username];
+    const virtualContestUrl = `/contest/${contestOptions.contestAlias}/virtual/`;
+
+    contestOptions.startDate = addSubtractDaysToDate(now, { days: -1 });
+    contestOptions.endDate = now;
```
Comment (heduenas) [inline cypress/integration/contest.spec.ts:132]: Can you try this to test virtual contests more deeply:
1. Create the (original) contest setting startDate = Now, endDate = now + 10 seconds
2. Right after creating the contest, make a submission using user 0.
3. Right after the contest ends, create a virtual copy of the contest using user 1.
4. Make a submission in the virtual contest using user 1.
5. After the virtual contest ends, verify that the scoreboard of the virtual contest shows both the submission of user 0 as well as the submission of user 1.

File: cypress/integration/contest.spec.ts [hunk 2]
```diff
@@ -105,16 +98,129 @@ describe('Contest Test', () => {
...
+      .should('have.attr', 'href')
+      .and('include', contestOptions.contestAlias);
+    contestPage.createClarificationUser(contestOptions, 'Question 1');
+    cy.logout();
+
+    contestPage.answerClarification(contestOptions, 'No');
+  });
+
+  it(
+    'Should create a contest and reviewing ranking contests when the scoreboard shows' +
```
Comment (heduenas) [inline cypress/integration/contest.spec.ts:191]: "reviewing ranking contests" -> "review contest ranking"?

File: cypress/support/pageObjects/loginPage.ts [hunk 3]
```diff
@@ -1,24 +1,23 @@
...
 
 export class LoginPage {
+  registerMultipleUsers(noOfUsers: number): LoginOptions[] {
+    const users: LoginOptions[] = [];
 
-    registerMultipleUsers(noOfUsers: number): LoginOptions[] {
-        const users: LoginOptions[] = [];
+    for (let i = 0; i < noOfUsers; i++) {
+      const userLoginOptions: LoginOptions = {
+        username: 'utGroup_' + uuid(),
```
Comment (heduenas) [inline cypress/support/pageObjects/loginPage.ts:10]: To make it easier when debugging these tests, can you rename the usernames to `'utGroup_user' + i + "_" + uuid()`

File: cypress/support/pageObjects/contestPage.ts [hunk 4]
```diff
@@ -6,173 +6,188 @@ import { ContestOptions, GroupOptions } from '../types';
...
-
-    generateContestOptions(): ContestOptions {
-        const now = new Date();
-
-        const contestOptions: ContestOptions = {
-            contestAlias: 'contest' + uuid().slice(0, 5),
-            description: 'Test Description',
-            startDate: addSubtractDaysToDate(now, { days: -1 }),
-            endDate: addSubtractDaysToDate(now, { days: 2 }),
-            showScoreboard: true,
```
Comment (heduenas) [inline cypress/support/pageObjects/contestPage.ts:150]: Can you add one test where the scoreboard is never shown, and one test where the scoreboard is shown only 50% of the time?

File: cypress/support/pageObjects/contestPage.ts [hunk 5]
```diff
@@ -6,173 +6,188 @@ import { ContestOptions, GroupOptions } from '../types';
...
-        cy.url().should('include', '/arena/' + contestOptions.contestAlias);
-
-        cy.get('a[href="#ranking"]').click();
-        cy.waitUntil(() =>
-            cy.get('[data-table-scoreboard]').should('be.visible')
-        );
-        cy.logout();
-    };
-
-    generateContestOptions(): ContestOptions {
```
Comment (heduenas) [inline cypress/support/pageObjects/contestPage.ts:142]: Can you add a test where "Different starts" is enabled?

File: cypress/support/pageObjects/loginPage.ts (hunk 3 above)
Comment (ShivamJhaa) [reply cypress/support/pageObjects/loginPage.ts:10]: Done.

File: cypress/integration/contest.spec.ts (hunk 2 above)
Comment (ShivamJhaa) [reply cypress/integration/contest.spec.ts:191]: Done.

File: cypress/integration/contest.spec.ts (hunk 1 above)
Comment (ShivamJhaa) [reply cypress/integration/contest.spec.ts:132]: Done.

File: cypress/support/pageObjects/contestPage.ts (hunk 4 above)
Comment (ShivamJhaa) [reply cypress/support/pageObjects/contestPage.ts:150]: Done.

File: cypress/support/pageObjects/contestPage.ts (hunk 5 above)
Comment (ShivamJhaa) [reply cypress/support/pageObjects/contestPage.ts:142]: Done.

**Comment by heduenas** [comment]
`Should create a contest when the scoreboard is shown half the time` is failing. It's probably flaky. I'm thinking that we can do this instead:
* Create one test where scoreboard visibility is 50%, startDate = now, endDate is 5h from now. Make a submission and it should show up in the scoreboard.
* Create another test where scoreboard visibility is 50%, startDate = 2.5h ago, endDate is 2.5h from now. Make a submission and it should not show up in the scoreboard.

Those tests will not be flaky.

Other than that your changes look good to me.

**Comment by ShivamJhaa** [comment]
> `Should create a contest when the scoreboard is shown half the time` is failing. It's probably flaky. I'm thinking that we can do this instead:
> 
> * Create one test where scoreboard visibility is 50%, startDate = now, endDate is 5h from now. Make a submission and it should show up in the scoreboard.
> * Create another test where scoreboard visibility is 50%, startDate = 2.5h ago, endDate is 2.5h from now. Make a submission and it should not show up in the scoreboard.
> 
> Those tests will not be flaky.
> 
> Other than that your changes look good to me.

The issue was not related to the test mentioned test, actually, the error was: sometimes when we make a submission inside a contest the solution gets accepted as soon as we hit the submit button and this results in "AC" status even before the "new" solution status. 
I just remove those check and now the test are passing.

**Review by heduenas — APPROVED** [review]

**Pull request merged**